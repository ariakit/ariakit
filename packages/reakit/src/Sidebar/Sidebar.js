import React from "react";
import PropTypes from "prop-types";
import { prop, ifProp } from "styled-tools";
import hoistNonReactStatics from "hoist-non-react-statics";
import styled from "../styled";
import as from "../as";
import Overlay from "../Overlay";

const Component = props => (
  <Overlay
    defaultSlide={props.align === "right" ? "left" : "right"}
    {...props}
  />
);

hoistNonReactStatics(Component, Overlay);

const Sidebar = styled(Component)`
  top: 0;
  height: 100vh;
  transform: none;
  overflow: auto;
  left: ${ifProp({ align: "right" }, "auto", 0)};
  right: ${ifProp({ align: "right" }, 0, "auto")};
  ${prop("theme.Sidebar")};
`;

Sidebar.propTypes = {
  align: PropTypes.oneOf(["left", "right"])
};

Sidebar.defaultProps = {
  align: "left",
  translateX: 0,
  translateY: 0
};

export default as("div")(Sidebar);
