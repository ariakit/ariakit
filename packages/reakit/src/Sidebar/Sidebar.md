By default, `Sidebar` is a fixed [Overlay](/components/overlay) that appears on the left side of the screen.

```jsx
import { Block, Button, Backdrop } from "reakit";

<Sidebar.Container>
  {sidebar => (
    <Block>
      <Button as={Sidebar.Show} {...sidebar}>
        Open sidebar
      </Button>
      <Backdrop fade as={Sidebar.Hide} {...sidebar} />
      <Sidebar slide {...sidebar}>Sidebar</Sidebar>
    </Block>
  )}
</Sidebar.Container>
```

You can change its position with the `align` prop:

```jsx
import { Block, Button, Backdrop } from "reakit";

<Sidebar.Container>
  {sidebar => (
    <Block>
      <Button as={Sidebar.Show} {...sidebar}>
        Open right sidebar
      </Button>
      <Backdrop fade as={Sidebar.Hide} {...sidebar} />
      <Sidebar slide align="right" {...sidebar}>Sidebar</Sidebar>
    </Block>
  )}
</Sidebar.Container>
```

This is usually combined with [Portal](/components/portal):

```jsx
import { Block, Button, Backdrop, Portal } from "reakit";

<Sidebar.Container>
  {sidebar => (
    <Block>
      <Button as={Sidebar.Show} {...sidebar}>
        Open sidebar
      </Button>
      <Backdrop as={[Portal, Sidebar.Hide]} {...sidebar} />
      <Sidebar as={Portal} {...sidebar}>Sidebar</Sidebar>
    </Block>
  )}
</Sidebar.Container>
```

You can also put it inside another wrapper component tweaking its `position` and `height` css properties:

```jsx
import { Block, Button, Backdrop } from "reakit";

<Sidebar.Container>
  {sidebar => (
    <Block relative overflow="hidden">
      <Button as={Sidebar.Show} {...sidebar}>
        Open sidebar
      </Button>
      <Backdrop absolute fade as={Sidebar.Hide} {...sidebar} />
      <Sidebar absolute height="100%" slide {...sidebar}>
        Sidebar
      </Sidebar>
    </Block>
  )}
</Sidebar.Container>
```