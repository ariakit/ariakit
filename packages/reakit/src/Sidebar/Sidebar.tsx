import * as React from "react";
import * as PropTypes from "prop-types";
import { theme, ifProp } from "styled-tools";
import hoistNonReactStatics from "hoist-non-react-statics";
import styled from "../styled";
import as from "../as";
import Overlay, { OverlayProps } from "../Overlay";

export interface SidebarProps extends OverlayProps {
  align?: "left" | "right";
}

const SidebarComponent = (props: SidebarProps) => (
  <Overlay
    defaultSlide={props.align === "right" ? "left" : "right"}
    {...props}
  />
);

hoistNonReactStatics(SidebarComponent, Overlay);

const Sidebar = styled(SidebarComponent)`
  top: 0;
  height: 100vh;
  transform: none;
  overflow: auto;
  left: ${ifProp({ align: "right" }, "auto", 0)};
  right: ${ifProp({ align: "right" }, 0, "auto")};
  ${theme("Sidebar")};
`;

// @ts-ignore
Sidebar.propTypes = {
  align: PropTypes.oneOf(["left", "right"])
};

Sidebar.defaultProps = {
  align: "left",
  translateX: 0,
  translateY: 0
};

export default as("div")(Sidebar);
