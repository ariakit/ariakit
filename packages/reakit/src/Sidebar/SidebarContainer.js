import OverlayContainer from "../Overlay/OverlayContainer";

const SidebarContainer = OverlayContainer;

export default SidebarContainer;
