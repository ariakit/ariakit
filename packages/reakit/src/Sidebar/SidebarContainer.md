```jsx
import { Block, Button, Sidebar } from "reakit";

<Sidebar.Container>
  {({ visible, show, hide, toggle }) => (
    <Block>
      <Button onClick={show}>Show</Button>
      <Sidebar visible={visible}>
        <Button onClick={hide}>Hide</Button>
      </Sidebar>
    </Block>
  )}
</Sidebar.Container>
```
