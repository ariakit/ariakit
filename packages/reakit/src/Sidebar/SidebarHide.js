import { theme } from "styled-tools";
import styled from "../styled";
import as from "../as";
import OverlayHide from "../Overlay/OverlayHide";

const SidebarHide = styled(OverlayHide)`
  ${theme("SidebarHide")};
`;

export default as("button")(SidebarHide);
