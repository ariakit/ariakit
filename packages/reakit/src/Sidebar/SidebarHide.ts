import { theme } from "styled-tools";
import styled from "../styled";
import OverlayHide, { OverlayHideProps } from "../Overlay/OverlayHide";

export interface SidebarHideProps extends OverlayHideProps {}

const SidebarHide = styled(OverlayHide)<SidebarHideProps>`
  ${theme("SidebarHide")};
`;

export default SidebarHide;
