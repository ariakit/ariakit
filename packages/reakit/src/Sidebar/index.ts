import Sidebar from "./Sidebar";
import SidebarContainer from "./SidebarContainer";
import SidebarToggle from "./SidebarToggle";
import SidebarShow from "./SidebarShow";
import SidebarHide from "./SidebarHide";

export * from "./Sidebar";
export * from "./SidebarContainer";
export * from "./SidebarToggle";
export * from "./SidebarShow";
export * from "./SidebarHide";

export default Object.assign(Sidebar, {
  Container: SidebarContainer,
  Toggle: SidebarToggle,
  Show: SidebarShow,
  Hide: SidebarHide
});
