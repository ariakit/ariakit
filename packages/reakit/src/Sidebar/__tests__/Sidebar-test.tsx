import * as React from "react";
import { render } from "react-testing-library";
import Sidebar from "../Sidebar";

test("html attrs", () => {
  const { getByTestId } = render(
    <Sidebar id="test" aria-label="test" data-testid="test" />
  );
  expect(getByTestId("test")).toHaveAttribute("data-testid", "test");
  expect(getByTestId("test")).toHaveAttribute("aria-label", "test");
});

test("styled", () => {
  const { container } = render(<Sidebar />);
  expect(container.firstChild).toMatchSnapshot();
});

test("styled visible", () => {
  const { container } = render(<Sidebar visible />);
  expect(container.firstChild).toMatchSnapshot();
});

test("align right", () => {
  const { container } = render(<Sidebar visible align="right" />);
  expect(container.firstChild).toMatchSnapshot();
});
