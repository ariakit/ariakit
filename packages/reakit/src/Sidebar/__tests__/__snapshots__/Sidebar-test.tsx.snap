// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`align right 1`] = `
.c3 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c3:focus:not(:focus-visible) {
  outline: none;
}

.c2 {
  -webkit-transform: translate3d(0px,0px,0px);
  -ms-transform: translate3d(0px,0px,0px);
  transform: translate3d(0px,0px,0px);
}

.c2[aria-hidden="true"] {
  pointer-events: none;
  display: none !important;
}

.c1 {
  position: fixed;
  z-index: 19900410;
  left: 50%;
  top: 50%;
}

.c0 {
  top: 0;
  height: 100vh;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
  overflow: auto;
  left: auto;
  right: 0;
}

<div
  aria-hidden="false"
  aria-modal="true"
  class="c0 c1 c2 c3"
  role="dialog"
/>
`;

exports[`styled 1`] = `
.c3 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c3:focus:not(:focus-visible) {
  outline: none;
}

.c2 {
  -webkit-transform: translate3d(0px,0px,0px);
  -ms-transform: translate3d(0px,0px,0px);
  transform: translate3d(0px,0px,0px);
}

.c2[aria-hidden="true"] {
  pointer-events: none;
  display: none !important;
}

.c1 {
  position: fixed;
  z-index: 19900410;
  left: 50%;
  top: 50%;
}

.c0 {
  top: 0;
  height: 100vh;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
  overflow: auto;
  left: 0;
  right: auto;
}

<div
  aria-hidden="true"
  aria-modal="true"
  class="c0 c1 c2 c3"
  hidden=""
  role="dialog"
/>
`;

exports[`styled visible 1`] = `
.c3 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c3:focus:not(:focus-visible) {
  outline: none;
}

.c2 {
  -webkit-transform: translate3d(0px,0px,0px);
  -ms-transform: translate3d(0px,0px,0px);
  transform: translate3d(0px,0px,0px);
}

.c2[aria-hidden="true"] {
  pointer-events: none;
  display: none !important;
}

.c1 {
  position: fixed;
  z-index: 19900410;
  left: 50%;
  top: 50%;
}

.c0 {
  top: 0;
  height: 100vh;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
  overflow: auto;
  left: 0;
  right: auto;
}

<div
  aria-hidden="false"
  aria-modal="true"
  class="c0 c1 c2 c3"
  role="dialog"
/>
`;
